### PR TITLE
Keep languages defined in .gitattributes

### LINKED ISSUE #21377: Feature suggestion #14528 (language detection override for new unknown ones) doesn't seem to be working
### Description

I just tested with Gitea 1.17.2, and the feature suggestion #14528 doesn't seem to be working even though I think the comments on the issue suggest it *should* be.

Basically, make a new repo, add some non-empty test file that ends in a new unused file extension like `.bla`, then add a `.gitattributes` line:
```
*.bla linguist-language=BlaLanguage
```
The expected result is that the lines of your `.bla` file get counted and show up as a new entry on the language bar of the repository labelled `BlaLanguage`. This is however not the case, it just seems to be ignored.

### Gitea Version

1.17.2

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

See https://try.gitea.io/blablablablab/blatest

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

It's someone else's instance

### Database

_No response_

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #21377

This marks all "defined" languages in the `.gitattributes` file so they are not removed if they are not of type `programming` or `markup`.

grafik

### CODE REVIEW

**Comment by KN4CK3R** [comment]
If you are ok with that I could update the test too.

**Comment by zeripath** [comment]
use a map[string]bool{} instead of a map[string]struct{}{} then your code becomes a lot simpler.

**Comment by zeripath** [comment]
I might also say `keepLanguage` is probably a bad name - perhaps call it `explicitIncludedLanguage` and add a comment on its definition to explain the intent.

**Comment by zeripath** [comment]
Now, Some questions about the intention. 

Is intention that we should only include the size of non-program non-markup files that are explicit defined in .gitattributes?

For example, we have a repo tree containing:

```
a.jpg
b.jpeg
.gitattributes
```

Now let us assume that enry detects  `*.jpg` as a "JPEG Image"  but not `*.jpeg`. As `.jpg` is not explicitly within the .gitattributes and it is not a markup or programming file the file size count is discarded. 

Then consider we add `.jpeg` to `.gitattributes` and mark as as "JPEG Image". This means its size is added to the file size for "JPEG Image" along with `a.jpg`. 

Now the file-size displayed includes `a.jpg` and `b.jpeg` but only `b.jpeg` is in the `.gitattributes`

Is that the intention? Or should it just be `b.jpeg`?

**Comment by KN4CK3R** [comment]
> use a map[string]bool{} instead of a map[string]struct{}{} then your code becomes a lot simpler.

We use `map[string]struct{}{}` as set currently. Now that we have generics we could build a proper set type.

> Is that the intention?

These groups are special cases. The intention of this PR is to support types enry does not detect. For me, adding a file type to `.gitattributes` indicates that I want to see this (group of) type(s) in the statistics.

**Comment by zeripath** [comment]
I may not have explained it clearly enough

.enry will detect certain files automatically as being of certain "language" with the type of "Data" or otherwise

.Assume that entry matches all .xls files as Excel data.

now assume you have a.xls, and b.xls and you only add b.xls to the gitattributes with a linguist-language tag of Excel.

The current code will sum the file size of both a.xls and b.xls. Is that intended?

**Comment by zeripath** [comment]
> We use `map[string]struct{}{}` as set currently. Now that we have generics we could build a proper set type.
> 
> >

i have always used map[string]bool for this. Ot is substantially clearer than struct and will use far less memory too

File: modules/git/repo_language_stats_gogit.go
Comment (delvh) [inline modules/git/repo_language_stats_gogit.go]: ~~If I see that correctly, you simply want a constant time check if you can keep languages, right?
Why not simplify it to `map[string]bool` then, which is both clearer and more performant?~~
EDIT: I noticed @zeripath s comments only now.

File: modules/git/repo_language_stats_gogit.go
Comment (delvh) [inline modules/git/repo_language_stats_gogit.go]: This call could then be simplified as well.

File: modules/git/repo_language_stats_nogogit.go
Comment (delvh) [inline modules/git/repo_language_stats_nogogit.go]: See above.

**Comment by KN4CK3R** [comment]
> i have always used map[string]bool for this. It is substantially clearer than struct and will use far less memory too

That's not true if [this](https://emersion.fr/blog/2017/sets-in-go/) is still valid today. Anyway, adding a set type now.

**Comment by lafriks** [comment]
> > We use `map[string]struct{}{}` as set currently. Now that we have generics we could build a proper set type.
> > >
> 
> i have always used map[string]bool for this. It is substantially clearer than struct and will use far less memory too

I don't think that's true, struct{}{} will use less memory than bool

**Comment by lafriks** [comment]
> > i have always used map[string]bool for this. It is substantially clearer than struct and will use far less memory too
> 
> That's not true if [this](https://emersion.fr/blog/2017/sets-in-go/) is still valid today. Anyway, adding a set type now.

Still valid:
```
name                        time/op
SetWithBoolValueWrite        3.81s ± 3%
SetWithStructValueWrite      3.54s ± 3%

name                        alloc/op
SetWithBoolValueWrite        885MB ± 0%
SetWithStructValueWrite      804MB ± 0%

name                        allocs/op
SetWithBoolValueWrite         614k ± 0%
SetWithStructValueWrite       614k ± 0%
```

**Comment by zeripath** [comment]
> > i have always used map[string]bool for this. It is substantially clearer than struct and will use far less memory too
> 
> That's not true if [this](https://emersion.fr/blog/2017/sets-in-go/) is still valid today. Anyway, adding a set type now.

 I stand corrected!

**Comment by zeripath** [comment]
But please answer the above question - is it intended that only the files matching the . gitattributes set linguist-language should be counted or is it that once one thing is matched in gitattributes all of that file type will be counted?

My suspicion is that we should be counting all programming language, all scripts and any files that match .gitattributes (but only those files). This is not what this PR does.

**Comment by KN4CK3R** [comment]
> and any files that match .gitattributes (but only those files). This is not what this PR does.

Repo structure:
```
/.gitattributes
/a.bar
/a.foo
/b.bar
/b.foo
/c.dummy
```

Content of `.gitattributes`:
```
a.foo linguist-language=FooLang
*.bar linguist-language=BarLang
```

Added logging to the `GetLanguageStats` method prints:
```
File: .gitattributes
File: a.bar
> has language: BarLang
File: a.foo
>  has language: FooLang
File: b.bar
> has language: BarLang
File: b.foo
File: c.dummy
```

So both `.bar` files are detected but only the `a.foo` file, as stated in the `.gitattributes`. I think that's exactly the behaviour you propose.

File: modules/git/repo_language_stats_gogit.go
Comment (Gusted) [inline modules/git/repo_language_stats_gogit.go]: Should this PR be blocked until #21408 is merged?

Comment (KN4CK3R) [reply modules/git/repo_language_stats_gogit.go]: I will update the (other) PR whichever gets merged first.

**Comment by KN4CK3R** [comment]
> will have both added to the `E-mail` size

I would say that's exactly what's intended? Enry puts all `.cpp, .c++, .cc,, .cp, .cxx, ...` in the language group `C++`. If you add `*.foobar linguist-language=C++` you are saying "and .foobar too please".

> There's also another problem:

I think that's out of scope for this PR. But as there is no issue for "ungroup a specific language" I don't think that's a real world problem at the moment. A possible solution could be a custom `*.ini linguist-language=EditorConfig x-linguist-leave-me-alone`.

File: modules/git/repo_language_stats_gogit.go
Comment (KN4CK3R) [reply modules/git/repo_language_stats_gogit.go]: done

**Comment by zeripath** [comment]
> > will have both added to the `E-mail` size
> 
> I would say that's exactly what's intended? Enry puts all `.cpp, .c++, .cc,, .cp, .cxx, ...` in the language group `C++`. If you add `*.foobar linguist-language=C++` you are saying "and .foobar too please".

You've still not understood what I'm saying,

* `E-mail` is `type.Data` and not a `type.Programming` or `type.Markup`.
* By default enry will map `.eml` and `.mbox` to `E-mail`
* So a repo with an .eml and .mbox file will have these sizes suppressed.
* If you add `*.eml linguist-language=E-mail` to .gitattributes you will enable the `E-mail`count to be shown
* But that count will now always contain .eml and .mbox file sizes
* Remove the .eml file and suddenly the size drops back to 0.

This is going to be confusing!

**Comment by KN4CK3R** [comment]
Still I would say it is intended behaviour because it does not change magically, the reason is a user action. For me it's a behaviour we just have to add to the documentation.

**Comment by ell1e** [comment]
> This is going to be confusing!

If my input is relevant at all (I'm just a user, although the one who requested this), I think the expectation would be that the `.gitattribute` overrides lead to no changes whatsoever to any files that aren't matched. Given that, I agree it might be unexpected and confusing

**Comment by zeripath** [comment]
@KN4CK3R 

```patch
diff --git a/modules/git/repo_language_stats_gogit.go b/modules/git/repo_language_stats_gogit.go
index cf2901266..5064ec7e2 100644
--- a/modules/git/repo_language_stats_gogit.go
+++ b/modules/git/repo_language_stats_gogit.go
@@ -12,7 +12,6 @@ import (
 	"strings"
 
 	"code.gitea.io/gitea/modules/analyze"
-	"code.gitea.io/gitea/modules/container"
 
 	"github.com/go-enry/go-enry/v2"
 	"github.com/go-git/go-git/v5"
@@ -45,10 +44,15 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 	checker, deferable := repo.CheckAttributeReader(commitID)
 	defer deferable()
 
+	// sizes contains the current calculated size of all files by language
 	sizes := make(map[string]int64)
-	// Languages explicit specifiyed with linguist-language or gitlab-language are always displayed
-	// even if they are not of type programming or markup.
-	explicitIncludedLanguage := make(container.Set[string])
+
+	// by default we will only count the sizes of programming languages or markup languages
+	// unless they are explicitly set using linguist-language
+	includedLanguage := map[string]bool{}
+	// or if there's only one language in the repository
+	firstExcludedLanguage, firstExcludedLanguageSize := "", int64(0)
+
 	err = tree.Files().ForEach(func(f *object.File) error {
 		if f.Size == 0 {
 			return nil
@@ -78,9 +82,8 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 					if len(group) != 0 {
 						language = group
 					}
-
+					// this language will always be added to the size
 					sizes[language] += f.Size
-					explicitIncludedLanguage.Add(language)
 					return nil
 				} else if language, has := attrs["gitlab-language"]; has && language != "unspecified" && language != "" {
 					// strip off a ? if present
@@ -93,9 +96,8 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 						if len(group) != 0 {
 							language = group
 						}
-
+						// this language will always be added to the size
 						sizes[language] += f.Size
-						explicitIncludedLanguage.Add(language)
 						return nil
 					}
 				}
@@ -129,7 +131,16 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 			language = group
 		}
 
-		sizes[language] += f.Size
+		included, checked := includedLanguage[language]
+		if !checked {
+			langtype := enry.GetLanguageType(language)
+			includedLanguage[language] = langtype != enry.Programming && langtype != enry.Markup
+		}
+		if included {
+			sizes[language] += f.Size
+		} else if len(sizes) == 0 && (firstExcludedLanguage == "" || firstExcludedLanguage == language) {
+			firstExcludedLanguageSize += f.Size
+		}
 
 		return nil
 	})
@@ -137,14 +148,9 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 		return nil, err
 	}
 
-	// filter special languages unless they are the only language
-	if len(sizes) > 1 {
-		for language := range sizes {
-			langtype := enry.GetLanguageType(language)
-			if langtype != enry.Programming && langtype != enry.Markup && !explicitIncludedLanguage.Contains(language) {
-				delete(sizes, language)
-			}
-		}
+	// If there are no included languages add the first excluded language
+	if len(sizes) == 0 && firstExcludedLanguage != "" {
+		sizes[firstExcludedLanguage] = firstExcludedLanguageSize
 	}
 
 	return sizes, nil
diff --git a/modules/git/repo_language_stats_nogogit.go b/modules/git/repo_language_stats_nogogit.go
index 8d982f487..c6a45262f 100644
--- a/modules/git/repo_language_stats_nogogit.go
+++ b/modules/git/repo_language_stats_nogogit.go
@@ -14,7 +14,6 @@ import (
 	"strings"
 
 	"code.gitea.io/gitea/modules/analyze"
-	"code.gitea.io/gitea/modules/container"
 	"code.gitea.io/gitea/modules/log"
 
 	"github.com/go-enry/go-enry/v2"
@@ -68,10 +67,16 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 
 	contentBuf := bytes.Buffer{}
 	var content []byte
+
+	// sizes contains the current calculated size of all files by language
 	sizes := make(map[string]int64)
-	// Languages explicit specifiyed with linguist-language or gitlab-language are always displayed
-	// even if they are not of type programming or markup.
-	explicitIncludedLanguage := make(container.Set[string])
+
+	// by default we will only count the sizes of programming languages or markup languages
+	// unless they are explicitly set using linguist-language
+	includedLanguage := map[string]bool{}
+	// or if there's only one language in the repository
+	firstExcludedLanguage, firstExcludedLanguageSize := "", int64(0)
+
 	for _, f := range entries {
 		select {
 		case <-repo.Ctx.Done():
@@ -112,7 +117,6 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 					}
 
 					sizes[language] += f.Size()
-					explicitIncludedLanguage.Add(language)
 					continue
 				} else if language, has := attrs["gitlab-language"]; has && language != "unspecified" && language != "" {
 					// strip off a ? if present
@@ -127,7 +131,6 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 						}
 
 						sizes[language] += f.Size()
-						explicitIncludedLanguage.Add(language)
 						continue
 					}
 				}
@@ -186,18 +189,22 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
 			language = group
 		}
 
-		sizes[language] += f.Size()
+		included, checked := includedLanguage[language]
+		if !checked {
+			langtype := enry.GetLanguageType(language)
+			includedLanguage[language] = langtype != enry.Programming && langtype != enry.Markup
+		}
+		if included {
+			sizes[language] += f.Size()
+		} else if len(sizes) == 0 && (firstExcludedLanguage == "" || firstExcludedLanguage == language) {
+			firstExcludedLanguageSize += f.Size()
+		}
 		continue
 	}
 
-	// filter special languages unless they are the only language
-	if len(sizes) > 1 {
-		for language := range sizes {
-			langtype := enry.GetLanguageType(language)
-			if langtype != enry.Programming && langtype != enry.Markup && !explicitIncludedLanguage.Contains(language) {
-				delete(sizes, language)
-			}
-		}
+	// If there are no included languages add the first excluded language
+	if len(sizes) == 0 && firstExcludedLanguage != "" {
+		sizes[firstExcludedLanguage] = firstExcludedLanguageSize
 	}
 
 	return sizes, nil
diff --git a/templates/code/searchform.tmpl b/templates/code/searchform.tmpl
index 26611cf4b..17d1944e1 100644
--- a/templates/code/searchform.tmpl
+++ b/templates/code/searchform.tmpl
@@ -1,14 +1,14 @@
 <form class="ui form ignore-dirty" style="max-width: 100%">
 	<div class="ui fluid action input">
-		<input name="q" value="{{.Keyword}}"{{if .CodeIndexerUnavailable }} disabled{{end}} placeholder="{{.locale.Tr "explore.search"}}…" autofocus>
-		<div class="ui dropdown selection tooltip{{if .CodeIndexerUnavailable }} disabled{{end}}" data-content="{{.locale.Tr "explore.search.type.tooltip"}}">
-			<input name="t" type="hidden" value="{{.queryType}}"{{if .CodeIndexerUnavailable }} disabled{{end}}>{{svg "octicon-triangle-down" 14 "dropdown icon"}}
+		<input name="q" value="{{.Keyword}}"{{if .CodeIndexerUnavailable}} disabled{{end}} placeholder="{{.locale.Tr "explore.search"}}…" autofocus>
+		<div class="ui dropdown selection tooltip{{if .CodeIndexerUnavailable}} disabled{{end}}" data-content="{{.locale.Tr "explore.search.type.tooltip"}}">
+			<input name="t" type="hidden" value="{{.queryType}}"{{if .CodeIndexerUnavailable}} disabled{{end}}>{{svg "octicon-triangle-down" 14 "dropdown icon"}}
 			<div class="text">{{.locale.Tr (printf "explore.search.%s" (or .queryType "fuzzy"))}}</div>
 			<div class="menu transition hidden" tabindex="-1" style="display: block !important;">
 				<div class="item tooltip" data-value="" data-content="{{.locale.Tr "explore.search.fuzzy.tooltip"}}">{{.locale.Tr "explore.search.fuzzy"}}</div>
 				<div class="item tooltip" data-value="match" data-content="{{.locale.Tr "explore.search.match.tooltip"}}">{{.locale.Tr "explore.search.match"}}</div>
 			</div>
 		</div>
-		<button class="ui primary button"{{if .CodeIndexerUnavailable }} disabled{{end}}>{{.locale.Tr "explore.search"}}</button>
+		<button class="ui primary button"{{if .CodeIndexerUnavailable}} disabled{{end}}>{{.locale.Tr "explore.search"}}</button>
 	</div>
 </form>
diff --git a/templates/user/code.tmpl b/templates/user/code.tmpl
index 730d4dc9e..ecd6aa566 100644
--- a/templates/user/code.tmpl
+++ b/templates/user/code.tmpl
@@ -5,13 +5,13 @@
 		{{template "code/searchform" .}}
 		<div class="ui divider"></div>
 		<div class="ui user list">
-			{{if .CodeIndexerUnavailable }}
+			{{if .CodeIndexerUnavailable}}
 				<div class="ui error message">
 					<p>{{$.locale.Tr "explore.code_search_unavailable"}}</p>
 				</div>
 			{{else if .SearchResults}}
 				<h3>
-					{{.locale.Tr "explore.code_search_results" (.Keyword|Escape) | Str2html }}
+					{{.locale.Tr "explore.code_search_results" (.Keyword|Escape) | Str2html}}
 				</h3>
 				{{template "code/searchresults" .}}
 			{{else if .Keyword}}

```

File: modules/git/repo_language_stats_gogit.go
Comment (lunny) [reply modules/git/repo_language_stats_gogit.go]: #21408 merged

**Comment by zeripath** [comment]
So one final question:

```go
						// group languages, such as Pug -> HTML; SCSS -> CSS
						group := enry.GetLanguageGroup(language)
						if len(group) != 0 {
							language = group
						}
```

Do we still want the grouping to occur? It's fine either way.

**Review by zeripath — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**
